### PR TITLE
[PATCH v2] Sched automatic worker group config

### PR DESCRIPTION
**Pull request by psavol**
Add config file options to disable unused automatic schedule groups. These are enabled by default as API spec requires it. By default, scheduler polls/checks if there are active queues in both "all" and "worker" groups (for a worker thread). If only "all" groups is used, disabling "worker" group improves performance a bit. The improvement is significant when burst size is small (as there are more schedule() call cycles per event). With a sched_pktio test case (2 stage pipeline) and burst size forced to 1, improvement was 4-7%.

This config can be later added to schedule config API and this implementation is reusable for that (config file option changes to API option, but other code remain the same).

Sched_pktio burst size change was needed to test the performance effect.

This needs to be rebased on top of #736 once it's merged.

### CODE REVIEW

File: test/performance/odp_sched_pktio.c [hunk 1]
```diff
@@ -534,6 +537,7 @@ static void print_usage(const char *progname)
 	       "  -i, --interface <name>    Packet IO interfaces (comma-separated, no spaces)\n"
 	       "  -c, --num_cpu <number>    Worker thread count. Default: 1\n"
 	       "  -q, --num_queue <number>  Number of pktio queues. Default: Worker thread count\n"
+	       "  -b, --burst <number>      Maximum number of events requested from scheduler. Default: 32\n"
```
Comment (Bill-Fischofer-Linaro) [inline test/performance/odp_sched_pktio.c]: It's not clear that this patch is related to the others in this PR. It seems to have nothing to do with scheduler groups. This seems a reasonable improvement for this test but shouldn't this be its own PR?

File: config/odp-linux-generic.conf [hunk 2]
```diff
@@ -100,4 +100,15 @@ sched_basic: {
 	burst_size_default = [ 32,  32,  32,  32,  32, 16,  8, 4]
 	burst_size_max     = [255, 255, 255, 255, 255, 16, 16, 8]
 
+	# Automatically updated schedule groups
+	#
+	# API specification defines that ODP_SCHED_GROUP_ALL,
+	# _WORKER and _CONTROL are updated automatically. These options can be
+	# used to disable these group when not used. Set value to 0 to disable
+	# a group. Performance may improve when unused groups are disabled.
```
Comment (Bill-Fischofer-Linaro) [inline config/odp-linux-generic.conf]: Threads are created as worker or control threads and queues are created in a specific scheduler group (default = `ODP_SCHED_GROUP_ALL`) so I'm not sure I understand the use cases envisioned here. What happens if `all = 0` is specified and `odp_queue_create()` is called to create a `SCHED` queue without an explicit scheduling group, so that by default it's assigned to group `ODP_SCHED_GROUP_ALL`?

Omitting the explicit worker and control groups might make sense if queues are either going to be in the default `all` group or explicitly named groups, but being able to disable the `all` group seems error-prone and it's not obvious under what conditions it would be useful.

File: platform/linux-generic/include/odp_schedule_if.h [hunk 3]
```diff
@@ -44,7 +53,7 @@ typedef void (*schedule_order_unlock_lock_fn_t)(void);
 typedef void (*schedule_order_lock_start_fn_t)(void);
 typedef void (*schedule_order_lock_wait_fn_t)(void);
 typedef uint32_t (*schedule_max_ordered_locks_fn_t)(void);
-typedef void (*schedule_save_context_fn_t)(uint32_t queue_index);
+typedef void (*schedule_config_fn_t)(schedule_config_t *config);
```
Comment (Bill-Fischofer-Linaro) [inline platform/linux-generic/include/odp_schedule_if.h]: Looks like `schedule_save_context_fn_t` is unused, so this deletion of a stale symbol should technically be a separate patch. This patch should just be adding `schedule_config_fn_t`. Minor point and probably not worrying about here.

File: test/performance/odp_sched_pktio.c (hunk 1 above)
Comment (psavol) [reply test/performance/odp_sched_pktio.c]: Cover letter says: "Sched_pktio burst size change was needed to test the performance effect."

File: config/odp-linux-generic.conf (hunk 2 above)
Comment (psavol) [reply config/odp-linux-generic.conf]: If user/application creates its own groups and places all queues in to those groups. "All" groups is not used or needed. User needs to be careful when changing configuration from the default.

File: platform/linux-generic/include/odp_schedule_if.h (hunk 3 above)
Comment (psavol) [reply platform/linux-generic/include/odp_schedule_if.h]: I think it would be wasteful (for review and git history) to create an extra patch for this removal of dead code while adding a new line next to it.

File: test/performance/odp_sched_pktio.c (hunk 1 above)
Comment (Bill-Fischofer-Linaro) [reply test/performance/odp_sched_pktio.c]: OK

File: config/odp-linux-generic.conf (hunk 2 above)
Comment (Bill-Fischofer-Linaro) [reply config/odp-linux-generic.conf]: Unless the application has complete control over all queue creation (which it does not, since ODP creates queues internally) it seems that allowing the `all` group to be removed is likely to cause problems. Is the performance gain so significant that this is worth it?

**Comment by muvarov** [comment]
Merged.
